addons/yatl/chat/chat.gd

```diff
@@ -1,163 +1,171 @@
+	else:
+		emit_signal("connection_failed")
 
 
 func chat_disconnect() -> void:
-	if !(__socket.get_peer(1).is_connected_to_host()):
+	if !__socket_connection || !__socket_connection.connected:
 		return
-
-	__socket.disconnect_from_host()
+		
+	__socket_connection.client.disconnect_from_host() # TODO: velop expose disconnect_from_host() to WebsocketConnection
+	yield(__socket_connection, "disconnected")
+	__client = null
+	__socket_connection = null
 
 
-func chat_auth(nick: String, token: String) -> void:
-	__socket.get_peer(1).set_write_mode(WebSocketPeer.WRITE_MODE_TEXT)
-	__socket_send("PASS " + ("" if token.begins_with("oauth:") else "oauth:") + token)
-	__socket_send("NICK " + nick.to_lower())
-	__socket_send("CAP REQ :twitch.tv/commands twitch.tv/tags twitch.tv/membership")
+func authenticate(nick: String, token: String) -> void:
+	if !__socket_connection || !__socket_connection.connected:
+		return
+	
+	__client.get_peer(1).set_write_mode(WebSocketPeer.WRITE_MODE_TEXT)
+	__client_put_message("PASS " + ("" if token.begins_with("oauth:") else "oauth:") + token)
+	__client_put_message("NICK " + nick.to_lower())
+	__client_put_message("CAP REQ :twitch.tv/commands twitch.tv/tags twitch.tv/membership")
 
 
 func join_channel(channel: String) -> void:
-	__socket_send("JOIN " + ("" if channel.begins_with("#") else "#") + channel.to_lower())
-#	__current_channel = channel
-
+	__client_put_message("JOIN " + ("" if channel.begins_with("#") else "#") + channel.to_lower().strip_edges())
 
-func leave_channel(channel: String = "") -> void:
-#	if __current_channel == "":
-#		return
-
-	__socket_send("PART " + ("" if channel.begins_with("#") else "#") + channel.to_lower())
-#	__current_channel = ""
-
-
-func send_message(message: String) -> void: #TODO: message queue
-#	if __current_channel == "":
-#		return
-	pass
-#	__socket_send("PRIVMSG " + ("" if __current_channel.begins_with("#") else "#") + __current_channel + " :" + message)
 
+func leave_current_channel() -> void:
+	if __current_channel.empty():
+		return
+	
+	__client_put_message("PART " + __current_channel)
 
-#Private methods
-func __socket_connection_closed(clean: bool) -> void:
-	print("disconnected, clean ", clean)
-	emit_signal("chat_disconnected", self, clean)
 
+func send_message(message: String) -> void:
+	__message_queue.append("PRIVMSG " + __current_channel + " :" + message)
 
-func __socket_connection_established(protocol: String) -> void:
-	print("connection established with protocol ", protocol)
-	emit_signal("chat_connected", self)
 
+# Private methods
 
-func __socket_connection_error() -> void:
-	print("connection error")
-	emit_signal("chat_connection_failed", self)
+func __on_socket_connection_disconnected():
+	emit_signal("disconnected")
 
 
-func __socket_data_received() -> void:
-	var data: PoolByteArray = __socket.get_peer(1).get_packet()
-	var data_string: String = data.get_string_from_utf8().strip_edges(false)
-	emit_signal("chat_raw_data_received", self, data_string)
```

Comment:
This is an interesting one to think about. What level of skill / required complexity are we assuming for the consumer of YATL?

I think that it is smart to remove this, but how simple do we want to make it?

Reply:
I definitely wouldn't want it to feel "dumbed down", and I want to offer a reasonable amount of flexibility, but I also don't want it to be bloated. I originally intended to bring this method back, but it seems it got lost in the transition (I rewrote all the relevant parts from scratch instead of porting it from the old code.) I'm not as sure about bringing this one back, because in most cases the data sent is already being handled by YATL and presented to the user in a better readable format than a weirdly structured IRC command.

Reply:
Totally agree. I don't think that there is a lot of point to returning the IRC command to the user. 

This conversation makes me wonder how much we handle from a Twitch standpoint. i.e Is YATL responsible to creating/managing/triggering commands, or is it simply to act as a wrapper for Twitch and the consumer is able to build on top of them?

Reply:
I'm against making YATL (at least the chat part) a full solution like that. To me it's a low-ish level wrapper around Twitch.